From 51aec6e9e048be6a50b5f90d7bfae4dbb42a137c Mon Sep 17 00:00:00 2001
From: Swee Aun Khor <sakhor@efinixinc.com>
Date: Thu, 16 Jul 2026 13:25:10 +0800
Subject: [PATCH] gpio: spinal-lib: make irq_chip instance local

Replace the global spinal_lib_gpio irq_chip with a per instance copy
stored in struct spinal_lib_gpio.

This avoids sharing irq_chip state between multiple GPIO controller
instances and aligns with the GPIO IRQ framework expectation that
irq_chip instances are associated with a specific gpio controller.

Signed-off-by: Swee Aun Khor <sakhor@efinixinc.com>
---
 drivers/gpio/gpio-spinal-lib.c | 36 ++++++++++++++++++++--------------
 1 file changed, 21 insertions(+), 15 deletions(-)

diff --git a/drivers/gpio/gpio-spinal-lib.c b/drivers/gpio/gpio-spinal-lib.c
index 57a1691e0..2f1e2136d 100644
--- a/drivers/gpio/gpio-spinal-lib.c
+++ b/drivers/gpio/gpio-spinal-lib.c
@@ -33,6 +33,7 @@ struct spinal_lib_gpio {
 	raw_spinlock_t		lock;
 	void __iomem		*base;
 	struct gpio_chip	gc;
+	struct irq_chip		irqchip;
 	unsigned long		enabled;
 	unsigned		trigger[MAX_GPIO];
 	unsigned int		irq_parent[MAX_GPIO];
@@ -41,6 +42,21 @@ struct spinal_lib_gpio {
 	u32 interrupt_mask;
 };
 
+static int spinal_lib_gpio_irq_set_type(struct irq_data *d, unsigned trigger);
+static void spinal_lib_gpio_irq_mask(struct irq_data *d);
+static void spinal_lib_gpio_irq_unmask(struct irq_data *d);
+static void spinal_lib_gpio_irq_enable(struct irq_data *d);
+static void spinal_lib_gpio_irq_disable(struct irq_data *d);
+
+static const struct irq_chip spinal_lib_gpio_irqchip_template = {
+	.name		= "spinal_lib_gpio-gpio",
+	.irq_set_type	= spinal_lib_gpio_irq_set_type,
+	.irq_mask	= spinal_lib_gpio_irq_mask,
+	.irq_unmask	= spinal_lib_gpio_irq_unmask,
+	.irq_enable	= spinal_lib_gpio_irq_enable,
+	.irq_disable	= spinal_lib_gpio_irq_disable,
+};
+
 static void spinal_lib_gpio_assign_bit(void __iomem *ptr, int offset, int value)
 {
 	// It's frustrating that we are not allowed to use the device atomics
@@ -210,15 +226,6 @@ static void spinal_lib_gpio_irq_disable(struct irq_data *d)
 	spinal_lib_gpio_set_ie(chip, offset);
 }
 
-static struct irq_chip spinal_lib_gpio_irqchip = {
-	.name		= "spinal_lib_gpio-gpio",
-	.irq_set_type	= spinal_lib_gpio_irq_set_type,
-	.irq_mask	= spinal_lib_gpio_irq_mask,
-	.irq_unmask	= spinal_lib_gpio_irq_unmask,
-	.irq_enable	= spinal_lib_gpio_irq_enable,
-	.irq_disable	= spinal_lib_gpio_irq_disable,
-};
-
 static void spinal_lib_gpio_irq_handler(struct irq_desc *desc)
 {
 	struct irq_chip *irqchip = irq_desc_get_chip(desc);
@@ -273,6 +280,7 @@ static int spinal_lib_gpio_probe(struct platform_device *pdev)
 	}
 
 	raw_spin_lock_init(&chip->lock);
+	chip->irqchip = spinal_lib_gpio_irqchip_template;
 	chip->gc.direction_input = spinal_lib_gpio_direction_input;
 	chip->gc.direction_output = spinal_lib_gpio_direction_output;
 	chip->gc.get_direction = spinal_lib_gpio_get_direction;
@@ -297,27 +305,25 @@ static int spinal_lib_gpio_probe(struct platform_device *pdev)
 	chip->interrupt_mask = 0xFFFFFFFF;
 	chip->enabled = 0;
 
-
-
-	ret = gpiochip_irqchip_add(&chip->gc, &spinal_lib_gpio_irqchip, 0, handle_level_irq, IRQ_TYPE_NONE);
+	ret = gpiochip_irqchip_add(&chip->gc, &chip->irqchip, 0, handle_level_irq, IRQ_TYPE_NONE);
 	if (ret) {
 		dev_err(dev, "could not add irqchip\n");
 		gpiochip_remove(&chip->gc);
 		return ret;
 	}
 
-	chip->gc.irq.num_parents = ngpio;
+	chip->gc.irq.num_parents = of_irq_count(node);
 	chip->gc.irq.parents = &chip->irq_parent[0];
 	chip->gc.irq.map = &chip->irq_parent[0];
 
-	for (gpio = 0; gpio < ngpio; ++gpio) {
+	for (gpio = 0; gpio < chip->gc.irq.num_parents; ++gpio) {
 		irq = platform_get_irq(pdev, gpio);
 		chip->irq_parent[gpio] = irq;
 		chip->self_ptr[gpio] = chip;
 		chip->trigger[gpio] = IRQ_TYPE_NONE;
 	}
 
-	for (gpio = 0; gpio < ngpio; ++gpio) {
+	for (gpio = 0; gpio < chip->gc.irq.num_parents; ++gpio) {
 		irq = chip->irq_parent[gpio];
 		if(irq < 0) continue;
 //		if (gpio > 1) continue;
-- 
2.43.0

